packages/mdc-toolbar/foundation.js

```diff
@@ -0,0 +1,225 @@
+    // Toolbar flexible behavior
+    // If the first row is flexible
+    this.hasFlexibleRow_ = false;
+    // If use the default behavior
+    this.useFlexDefaultBehavior_ = false;
+  }
+
+  init() {
+    this.fixedAllRow_ = this.adapter_.hasClass(MDCToolbarFoundation.cssClasses.FIXED);
+    this.fixedLastrow_ = this.adapter_.hasClass(MDCToolbarFoundation.cssClasses.FIXED_LASTROW);
+    this.hasFlexibleRow_ = this.adapter_.hasClass(MDCToolbarFoundation.cssClasses.TOOLBAR_ROW_FLEXIBLE);
+    if (this.hasFlexibleRow_) {
+      this.useFlexDefaultBehavior_ = this.adapter_.hasClass(MDCToolbarFoundation.cssClasses.FLEXIBLE_DEFAULT_BEHAVIOR);
+    }
+    this.initKeyRatio_();
+    this.setKeyHeights_();
+    this.adapter_.addClass(MDCToolbarFoundation.cssClasses.FIXED_UPGRADE);
+    this.adapter_.registerResizeHandler(this.resizeHandler_);
+    this.adapter_.registerScrollHandler(this.scrollHandler_);
+  }
+
+  destroy() {
+    this.adapter_.deregisterResizeHandler(this.resizeHandler_);
+    this.adapter_.deregisterScrollHandler(this.scrollHandler_);
+  }
+
+  updateAdjustElementStyles() {
+    if (this.fixedLastrow_ || this.fixedAllRow_) {
+      this.adapter_.setStyleForFixedAdjustElement('marginTop', `${this.calculations_.toolbarHeight}px`);
+    }
+  }
+
+  getFlexibleExpansionRatio_(scrollTop) {
+    const delta = 0.0001;
+    return Math.max(0, 1 - scrollTop / (this.calculations_.flexibleExpansionHeight + delta));
+  }
+
+  checkRowHeight_() {
+    cancelAnimationFrame(this.checkRowHeightFrame_);
+    this.checkRowHeightFrame_ = requestAnimationFrame(() => this.setKeyHeights_());
+  }
+
+  setKeyHeights_() {
+    const newToolbarRowHeight = this.getRowHeight_();
+    if (newToolbarRowHeight !== this.calculations_.toolbarRowHeight) {
+      this.calculations_.toolbarRowHeight = newToolbarRowHeight;
+      this.calculations_.toolbarHeight = this.calculations_.toolbarRatio * this.calculations_.toolbarRowHeight;
+      this.calculations_.flexibleExpansionHeight =
+        this.calculations_.flexibleExpansionRatio * this.calculations_.toolbarRowHeight;
+      this.calculations_.maxTranslateYDistance =
+        this.calculations_.maxTranslateYRatio * this.calculations_.toolbarRowHeight;
+      this.calculations_.scrollTheshold =
+        this.calculations_.scrollThesholdRatio * this.calculations_.toolbarRowHeight;
+      this.updateAdjustElementStyles();
+      this.updateToolbarStyles_();
+    }
+  }
+
+  updateToolbarStyles_() {
+    cancelAnimationFrame(this.scrollFrame_);
+    this.scrollFrame_ = requestAnimationFrame(() => {
+      const scrollTop = this.adapter_.getViewportScrollY();
+      const hasScrolledOutOfTheshold = this.scrolledOutOfTheshold_(scrollTop);
+
+      if (hasScrolledOutOfTheshold && this.executedLastChange_) {
+        return;
+      }
+
+      const flexibleExpansionRatio = this.getFlexibleExpansionRatio_(scrollTop);
+      this.adapter_.notifyChange({flexibleExpansionRatio: flexibleExpansionRatio});
```

Comment:
this is a tiny nit, but perhaps we should call `notifyChange` _after_ all of the updates have been made? This way, if the user wants to query the dimensions of the header or any of its children, it will have the most up-to-date dimensions.